pkg/expression/builtin_time.go

```diff
@@ -3567,7 +3567,7 @@ func (c *addSubDateFunctionClass) getFunction(ctx sessionctx.Context, args []Exp
 			resultEvalTp = types.ETDatetime
 		}
 	} else if dateEvalTp == types.ETDuration {
-		if types.IsDateUnit(unit) && unit != "DAY_MICROSECOND" {
+		if types.IsDateUnit(unit) && unit != "DAY_MICROSECOND" && unit != "DAY" {
```

Comment:
Please check other types can also be handled correctly.

When we execute `select date_add(o, interval "1 1" day_hour) from test_interval;` in mysql and tidb. We will get the following results respectively.
```sql
# mysql
mysql> select date_add(o, interval "1 1" day_hour) from test_interval;
Field   1:  `date_add(o, interval "1 1" day_hour)`
Catalog:    `def`
Database:   ``
Table:      ``
Org_table:  ``
Type:       TIME
Collation:  binary (63)
Length:     10
Max_length: 8
Decimals:   0
Flags:      BINARY


+--------------------------------------+
| date_add(o, interval "1 1" day_hour) |
+--------------------------------------+
| 26:02:03                             |
+--------------------------------------+

# tidb
mysql> select date_add(o, interval "1 1" day_hour) from test_interval;
Field   1:  `date_add(o, interval "1 1" day_hour)`
Catalog:    `def`
Database:   ``
Table:      ``
Org_table:  ``
Type:       DATETIME
Collation:  binary (63)
Length:     19
Max_length: 19
Decimals:   0
Flags:      BINARY 


+--------------------------------------+
| date_add(o, interval "1 1" day_hour) |
+--------------------------------------+
| 2023-10-27 02:02:03                  |
+--------------------------------------+
```

Interval with year type also lead to different return type. Please refer to this [table](https://dev.mysql.com/doc/refman/8.0/en/expressions.html#temporal-intervals:~:text=Table%C2%A09.2%C2%A0Temporal%20Interval%20Expression%20and%20Unit%20Arguments) and ensure all interval types will return correct types :)

Reply:
@xzhangxian1008 
We investigated the return value type in MySQL 8.2. The results showed that no modifications to the code were necessary.
According to the [MySQL documentation](https://dev.mysql.com/doc/refman/8.0/en/date-and-time-functions.html#function_date-add), if the date_add and date_sub arguments are given the TIME type, the return value is determined according to the following rules.

> (MySQL 8.0.28 and later:) TIME if the date argument is a TIME value and the calculations involve only HOURS, MINUTES, and SECONDS parts (that is, no date parts).

> DATETIME if the first argument is a DATETIME (or TIMESTAMP) value, or if the first argument is a DATE and the unit value uses HOURS, MINUTES, or SECONDS, or if the first argument is of type TIME and the unit value uses YEAR, MONTH, or DAY.

Existing implementations followed this rule. Therefore, there is no need to modify the code. An integration test has been added to verify this, but this PR can be closed without merging. Thank you for your review.